### DISCUSSION TITLE
Request for example of ensemble (e.g. average) of same forecaster with different random starting weights

### DISCUSSION THREAD

**Discussion by ericjb** [alan-turing-institute/sktime#5823] [Q&A]
I would like a pipeline that uses make_reduction() with an ML model (e.g. MLP via MLPRegressor), where the pipeline is given a parameter n_repeats, and the MLP fit is performed that many times (using different starting random weights). A forecast could be set to be an ensemble - say an average - of the n_repeats forecasts. Thanks for any pointers to existing examples or suggestions.

**Comment by fkiraly** [comment]
I think you can do this with `BaggingForecaster`. If you use `make_reduction`, you can also try using `BaggingRegressor` inside `make_reduction`, but this will of course only work for reduction forecasters.

**Reply by ericjb** [reply]
Thanks for the quick reply FK but bagging is not what I had in mind. I am trying to re-create R's nnetar forecasting function (from forecast::nnetar - by Hyndman and Casares - see https://www.rdocumentation.org/packages/forecast/versions/8.21.1/topics/nnetar). I think it should be possible to create a custom forecaster using sktime building blocks and sklearn's MLPRegressor to do this. I am still an sktime newbie but I will keep plugging away at this.

**Reply by fkiraly** [reply]
Oh, I see - you want `EnsembleForecaster` then.
(for fitting non-uniform weights, `AutoEnsembleForecaster`)

**Reply by fkiraly** [reply]
hm, might be non-intuitive, since `n_repeats` is not a tuneable parameter - so you have to create as many copies of the forecaster as you want.

Ways I can think of to get this with `n_repeats` or equivalent:

* in `make_reduction` use `BaggingRegressor`, and sample rows/columns *without* replacement at 1.0 proba. This will work only for reducers. though
* use `BaggingForecaster` with a `SplitterBootstrapTransformer`, a trivial splitter (e.g., `CutofSplitter`) with `Repeat` wrapped. This is a bit on the side of "too many components" though, as you plug in many components that are "trivial"

What I think would be a better "end state" is that `BaggingForecaster` does "trivial bagging" as a default, i.e., just pass on the full data repeatedly, instead of STL, possibly with an `n_repeats` param. And STL can be plugged in to get STL bootstrap.

What do you think?

**Reply by ericjb** [reply]
Hacking the example in the documentation for EnsembleForecaster, 
for the n_repeats issue I was thinking of something like

```
name = ["model-" + str(i) for i in range(n_repeats)]
forecasters = [
    ("trend", PolynomialTrendForecaster()),
]
for i in range(n_repeats):
    forecasters.append(name[i], NaiveForecaster())
forecaster = EnsembleForecaster(forecasters=forecasters, aggfunc='mean')
```

For the specific goal of the nnetar-like functionality
I would need to do some additional steps related to
TransformedTargetForecaster, scaling, make_reduction, MLPRegressor

Thoughts?

**Reply by fkiraly** [reply]
> Hacking the example in the documentation for EnsembleForecaster,

Yes, that's what I meat with "number of repeats" not being available as a parameter, and "too many components".
But it would probably work.

What do you think of the suggestions to simplify or extend some of the components?

**Reply by ericjb** [reply]
To be honest, I didn't completely understand your suggestions, or why they would lead to a simpler solution than my suggestion, at least in terms of programmer time. (my time)  I suspect you have a deeper understanding that n_repeats copies of the forecaster is inefficient/wasteful. In the meantime, the following is my intermediate progress. It works but I still need to package it up into a custom estimator that takes parameters.

```
n_repeats = 20
winlen = 7
name = ["model-" + str(i) for i in range(n_repeats)]
rand = np.random.randint(1, 1000, size=n_repeats)
forecasters = []
for i in range(n_repeats):
    regressor = MLPRegressor(random_state=rand[i], max_iter=500)
    forecasters.append((name[i], make_reduction(regressor, window_length=winlen, strategy="recursive")))
forecasterEnsemble = EnsembleForecaster(forecasters)
forecaster = StandardScaler() * forecasterEnsemble
forecaster.fit(y)
fh = ForecastingHorizon(np.arange(1,13), is_relative=True)
y_pred = forecaster.predict(fh)
plot_series(y, y_pred, labels=["y", "y_pred"])
```

**Reply by ericjb** [reply]
In the code above, after the statement forecaster.fit(y), is there a way to retrieve the fitted values for the training period? I would like to plot them alongside y. 

I tried "predicting" for fh = ForecastingHorizon(y.index, is_relative=False) but that generated the error 
NotImplementedError: Generating in-sample predictions is not yet implemented for RecursiveTabularRegressionForecaster.

Thanks for any pointers.

**Reply by fkiraly** [reply]
>  or why they would lead to a simpler solution than my suggestion, at least in terms of programmer time. (my time)

The idea was to minimize boilerplate, while also not making the syntax not complicated. I was just thinking back and forth what might be the easiest here, so your input is very much appreciated.

> I tried "predicting" for `fh = ForecastingHorizon(y.index, is_relative=False) `

Yes, that would be the usual way. Indeed in-sample forecasts are not implemented for this estimator.

**Reply by fkiraly** [reply]
you can search for forecasters - or ask a given one whether it has the capability - via the tag `capability:insample` (for search, use `registry.all_estimators`)

**Comment by fkiraly** [comment]
Perhaps this is helpful, @ericjb: https://github.com/sktime/sktime/pull/5854

I added a bootstrap transformer which just repeats the series.
If you use this in `BaggingForecaster`, it should result in "repeat and average".

This is how I imagine it working, I think this does the same as your code:

```python
n_repeats = 20
winlen = 7

regressor = MLPRegressor(max_iter=500)
forecaster = make_reduction(regressor, window_length=winlen, strategy="recursive"))

ensemble = BaggingForecaster(forecaster, RepeatBootstrapTransformer(n_repeats))

forecaster = StandardScaler() * ensemble

forecaster.fit(y)
fh = ForecastingHorizon(np.arange(1,13), is_relative=True)
y_pred = forecaster.predict(fh)
plot_series(y, y_pred, labels=["y", "y_pred"])
```

**Reply by ericjb** [reply]
Thanks for continuing to think about this @fkiraly. 
I downloaded the repeat-bootstrap branch and tried to run your code but it threw an error. Here's what I did

```
from sktime.forecasting.compose import BaggingForecaster
from sktime.transformations.bootstrap import RepeatBootstrapTransformer
from sktime.datasets import load_airline
y = load_airline()

n_repeats = 20
winlen = 7

regressor = MLPRegressor(max_iter=500)
forecaster1 = make_reduction(regressor, window_length=winlen, strategy="recursive")

ensemble = BaggingForecaster(forecaster1, RepeatBootstrapTransformer(n_repeats))

forecaster2 = StandardScaler() * ensemble

forecaster2.fit(y)
fhFK = ForecastingHorizon(np.arange(1,13), is_relative=True)
y_predFK = forecaster2.predict(fhFK)
plot_series(y, y_predFK, labels=["y", "y_predFK"])`
```

And here's the error

{
	"name": "TypeError",
	"message": "bootstrap_transformer in BaggingForecaster should be a Transformer that takes as input a Series and output a Panel.",
	"stack": "---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/tmp/ipykernel_53729/3206211541.py in <cell line: 16>()
     14 forecaster2 = StandardScaler() * ensemble
     15 
---> 16 forecaster2.fit(y)
     17 fhFK = ForecastingHorizon(np.arange(1,13), is_relative=True)
     18 y_predFK = forecaster2.predict(fhFK)

/usr/local/share/python/env/python_env_pycaret/lib/python3.10/site-packages/sktime/forecasting/base/_base.py in fit(self, y, X, fh)
    378         # we call the ordinary _fit if no looping/vectorization needed
    379         if not vectorization_needed:
--> 380             self._fit(y=y_inner, X=X_inner, fh=fh)
    381         else:
    382             # otherwise we call the vectorized version of fit

/usr/local/share/python/env/python_env_pycaret/lib/python3.10/site-packages/sktime/forecasting/compose/_pipeline.py in _fit(self, y, X, fh)
   1007         # fit forecaster
   1008         f = self.forecaster_
-> 1009         f.fit(y=yt, X=X, fh=fh)
   1010 
   1011         # transform post

/usr/local/share/python/env/python_env_pycaret/lib/python3.10/site-packages/sktime/forecasting/base/_base.py in fit(self, y, X, fh)
    378         # we call the ordinary _fit if no looping/vectorization needed
    379         if not vectorization_needed:
--> 380             self._fit(y=y_inner, X=X_inner, fh=fh)
    381         else:
    382             # otherwise we call the vectorized version of fit

/usr/local/share/python/env/python_env_pycaret/lib/python3.10/site-packages/sktime/forecasting/compose/_bagging.py in _fit(self, y, X, fh)
    180             and not isinstance(self.bootstrap_transformer_, BaseTransformer)
    181         ):
--> 182             raise TypeError(
    183                 \"bootstrap_transformer in BaggingForecaster should be a Transformer \"
    184                 \"that takes as input a Series and output a Panel.\"

TypeError: bootstrap_transformer in BaggingForecaster should be a Transformer that takes as input a Series and output a Panel."
}

**Reply by fkiraly** [reply]
Oh, sorry - if you are doing this with the current `BaggingForecaster`, you need to pass the transformer as the first argument.

I was already, apologies for the confusion, thinking about improving intuitiveness of `BaggingForecaster`, and when I said "how I imagine this working" it included alterations to its signature:

* switching the arguments, have `forecaster` first and `bootstrap_transformer` second
* have `RepeatBootstrapTransfomer` as default if the latter is not passed
    * this would also remove the `statsmodels` dependency for the default params
* if `int` is passed, interpret as `n_repeats`

Currently working code, on the branch, without these signature alterations (which should go in a separate PR), would be:

```python
n_repeats = 20
winlen = 7

regressor = MLPRegressor(max_iter=500)
forecaster = make_reduction(regressor, window_length=winlen, strategy="recursive"))

ensemble = BaggingForecaster(RepeatBootstrapTransformer(n_repeats), forecaster)

forecaster = StandardScaler() * ensemble

forecaster.fit(y)
fh = ForecastingHorizon(np.arange(1,13), is_relative=True)
y_pred = forecaster.predict(fh)
plot_series(y, y_pred, labels=["y", "y_pred"])
```

**Reply by ericjb** [reply]
Hi @fkiraly: Your modified code using BaggingForecaster and RepeatBootstrapTransformer runs with no errors. Thanks
Some comments and questions:

1. **Execution Time**: I compared execution time for my approach to yours (using a large number of repeats) and they are essentially the same.
2. **Reproducibility**: I tested reproducibility via  `regressor = MLPRegressor(max_iter=500, random_state=123).`   I then compared results for different settings of n_repeats. They did not change. This seems to imply that the same starting weights are used for each of the fits, so instead of n_repeats different fits you are getting the same fit, repeated n_repeats times. Is there a different way to get reproducibility or does this invalidate your approach?  (My approach handles the reproducibility directly by having n_repeats different regressors, each with a different setting for random_state.)
3. **In-sample vs Out-of-Sample** A few days ago I raised the issue of how to retrieve the fitted values for the training period. This is an important issue. For example, it is normal to compare accuracy in the training (in-sample) period with accuracy in the test (out-of-sample) period, to test for overfitting. I was wondering if there is some kind of workaround possible for this, for example by supplying windows from the training period as if they are "new inputs" for the forecaster. 
4. **make_reduction and seasonality** The R forecast::nnetar() function takes arguments 'p' and 'P' representing number of non-seasonal lags and number of seasonal lags, respectively. Suppose the time series y has periodicity m (say m=12 for a monthly series), and p=3 and P=2. Then the windows that need to be built for training do not represent consecutive observations from y, but would be of the form [y_{t-1}, y_{t-2}, y_{t-3}, y_{t-12}, y_{t-24}]. Does make_reduction support this type of subseries windows?

Thanks

**Reply by fkiraly** [reply]
> Is there a different way to get reproducibility or does this invalidate your approach? 

With current logic, it is not easily possible, but once https://github.com/sktime/skbase/pull/268 gets merged in `scikit-base`, you could add logic that internally generates dependent random seeds.

Your current approach works, though I do not see it as the best end state since (a) you have to set the random seeds manually, which is cumbersome and (b) manual setting by a simpl rule (rather than hash iteration) does not guarantee unbiased independent sampling of seeds.

> retrieve the fitted values for the training period.

If you mean, for the recursive reduction forecastere, there is a new prototype - which suffers from some runtime issues - `RecursiveReductionForecaster`, which can make in-sample forecasts. The old implementation does not offer this functionality, but it could be added.

> representing number of non-seasonal lags and number of seasonal lags

Not out of the box, but you can use the `Lag` estimator to produce arbitrary lags. There also has been a discussion on how to integrate arbitrary lags with `make_reduction`, e.g., passing a list of lags instead of just a window size. However, someone needs to commit to a small project to implement this.

**Reply by ericjb** [reply]
1. **Reproducibility** I created a custom estimator that supports reproducibility. Not cumbersome at all when I use it. Here's the initial __init__() function, which is missing a lot of nnetar's functionality. You can see where the seed is used. I think it's ok.
```
def __init__(self, p, P=0, size=None, repeats=1, xreg=None, lmbda=None, model=None, subset=None,
                 scale_inputs=True, seed=None, max_iter=500):

        self.p = p
        self.P = P
        self.size = size
        self.repeats = repeats
        self.scale_inputs = scale_inputs
        self.seed = seed
        self.max_iter = max_iter
        self.forecaster_ = None

        # ignored at this time (TBD xreg, lmbda, model, subset)
        self.xreg = xreg
        self.lmbda = lmbda
        self.model = model
        self.subset = subset
        
        # leave this as is
        super().__init__()

        # parameter checking logic (if applicable) should happen here
        self._size = size if size is not None else int(np.ceil((p + P) / 2)) + 1

        if self.seed is not None:
            np.random.seed(self.seed)
        rand = np.random.randint(1, 1000, size=self.repeats)

        name = ["model-" + str(i) for i in range(self.repeats)]
        forecasters = []
        for i in range(self.repeats):
            regressor = MLPRegressor(hidden_layer_sizes = np.array([self._size]),
                                     random_state=rand[i], max_iter=self.max_iter)
            forecasters.append((name[i],
                                make_reduction(regressor,
                                  window_length=self.p+self.P,strategy="recursive")))

        forecaster_ensemble = EnsembleForecaster(forecasters)
        if self.scale_inputs:
            self.forecaster = StandardScaler() * forecaster_ensemble
        else:
            self.forecaster = forecaster_ensemble

```

2. **In-sample forecasts** This sounds very interesting. How do I get access to the prototype RecursiveReductionForecaster? I would like to try it out.
3. **make-reduction small project** [a bit off topic] I will be teaching a course on time series and forecasting in the second semester. I thought of allowing (actually encouraging) students to contribute to sktime instead of the usual course projects. These are business-with-data-science students rather than computer science students, so I am not sure they have the necessary programming level, but still ... Happy to get suggestions for candidate projects. Can also be QA/QC/documentation projects.

**Reply by fkiraly** [reply]
> I created a custom estimator that supports reproducibility

Do you want to contribute that? Given that it looks almost complete. I believe you can save yourself some work by using `_DelegatedForecaster`.

> This sounds very interesting. How do I get access to the prototype `RecursiveReductionForecaster`? I would like to try it out.

Import from `sktime.forecasting.compose._reduce` directly. Issue tracking feedback and input is here: https://github.com/sktime/sktime/issues/3224

> I thought of allowing (actually encouraging) students to contribute to sktime instead of the usual course projects.

That would be really great! We have a mentoring programme, for academic projects you typically pair up a university mentor with an sktime mentor, see details:

https://github.com/sktime/mentoring

summer cohort starts onboarding Apr/May

Project ideas will be updated fo 2024 shortly, roadmap 2023-2024 (updated summer) is here: https://github.com/sktime/sktime/issues/4691

**Reply by ericjb** [reply]
Hi @fkiraly: I modified my forecaster to use RecursiveReductionForecaster() and it gives me the in-sample forecasts. Thanks! 🙏

**Comment by ericjb** [comment]
In an earlier entry you wrote:

_Not out of the box, but you can use the Lag estimator to produce arbitrary lags. There also has been a discussion on how to integrate arbitrary lags with make_reduction, e.g., passing a list of lags instead of just a window size. However, someone needs to commit to a small project to implement this._

Can you point me to the discussion related to "passing a list of lags instead of just a window size" ? I might consider this project (if it is small enough 🙂)

**Reply by fkiraly** [reply]
Oh, of course!

An ahead warning though, changes to `make_reduction` may be painful since it is organically grown, and we are working on structuing it better internally. Here are two threads reated to reduction forecasters:

* tracking issue for work items: https://github.com/sktime/sktime/issues/5736
* user discussion issue re features and rework: https://github.com/sktime/sktime/issues/3224

The types of contributions that would be very much appreciated would be:

* picking up one of the things in the tracking issue - not trying to do everything at once (not advised...)
* looking at the new reducer prototypes and cutting down on runtime (or making other improvements)
